apps/src/p5lab/spritelab/commands/validationCommands.js

```diff
@@ -57,10 +64,583 @@ export const commands = {
+    let idArray = [];
+
+    // Only check for values that are new this frame.
+    for (let i = this.previous.eventLogLength; i < this.eventLog.length; i++) {
+      if (
+        // Check for each event type that includes sprite ids (NOT time or key events).
+        this.eventLog[i].includes('whenClick: ') ||
+        this.eventLog[i].includes('whileClick: ') ||
+        this.eventLog[i].includes('whenTouch: ') ||
+        this.eventLog[i].includes('whileTouch: ') ||
+        this.eventLog[i].includes('spriteCreated: ')
+      ) {
+        // Use .concat because it's possible for multiple events to be logged in the same frame.
+        idArray = idArray.concat(
+          // Take whatever was in the current eventLog entry...
+          this.eventLog[i]
+            // ...remove the spaces and create an array...
+            .split(' ')
+            // ...convert each string in the array to a number...
+            .map(Number)
+            // ...remove NaN values. (ex. 'whenClick: 0' results in [Nan, 0] above)
+            .filter(function(value) {
+              return !Number.isNaN(value);
+            })
+        );
+      }
+    }
+    return idArray;
+  },
+
+  // Used in levels (typically first frame only) to create an ordered list of success criteria
+  addCriteria(predicate, feedback) {
+    if (typeof predicate === 'function' && typeof feedback === 'string') {
+      this.criteria.push(new criteria(predicate, feedback));
+    }
+  },
+
+  // Used in levels to override default validation timing.
+  setEarlyTime(frames) {
+    this.validationTimes.early = frames;
+  },
+  setWaitTime(frames) {
+    this.validationTimes.wait = frames;
+  },
+  setDelayTime(frames) {
+    this.validationTimes.delay = frames;
+  },
+
+  // Used in levels (typically every frame) to validate based on all criteria
+  validate() {
+    // Get the current (ie. previous frame) pass/fail state prior to validation
+    const state = commands.getPassState(this.criteria);
+
+    const barWidth =
+      this.currentFrame() * commands.calculateBarScale(this.validationTimes);
+    drawUtils.validationBar(this.p5, barWidth, state, {});
+
+    // Check all criteria and update the completion status of each.
+    if (this.currentFrame() <= this.validationTimes.wait) {
+      commands.checkAllCriteria(this.criteria);
+    } else {
+      // If "wait time" is over, determine if student passes or fails.
+      var results = {};
+      if (state === 'fail') {
+        console.log(this.criteria);
+        results = {
+          state: 'failed',
+          feedback: commands.reportFailure(this.criteria)
+        };
+      } else if (
+        this.currentFrame() >=
+        this.validationTimes.wait + this.validationTimes.delay
+      ) {
+        results = {
+          state: 'succeeded',
+          feedback: commands.reportSuccess(this.criteria)
+        };
+      }
+      return results;
+    }
+    commands.updatePrevious.call(this);
```

Comment:
Slightly confused.  I think this is storing the current state which will become the "previous" state next time?  The naming here sounds a bit like it's updating something related to the last frame, but I don't think that's true at the time of calling...

Reply:
Renamed and added comments.